Changed 3 modules and got  java.lang.StackOverflowError

I've changed 3 modules and got such a problem. Could @ismaeldivita you take a look?
```
> Task :changedModules FAILED
FAILURE: Build failed with an exception.
* What went wrong:
Execution failed for task ':changedModules'.
> java.lang.StackOverflowError (no error message)
* Try:
Run with --info or --debug option to get more log output. Run with --scan to get full insights.
* Exception is:
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':changedModules'.
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.lambda$executeIfValid$1(ExecuteActionsTaskExecuter.java:207)
	at org.gradle.internal.Try$Failure.ifSuccessfulOrElse(Try.java:263)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeIfValid(ExecuteActionsTaskExecuter.java:205)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:186)
	at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:114)
	at org.gradle.api.internal.tasks.execution.FinalizePropertiesTaskExecuter.execute(FinalizePropertiesTaskExecuter.java:46)
	at org.gradle.api.internal.tasks.execution.ResolveTaskExecutionModeExecuter.execute(ResolveTaskExecutionModeExecuter.java:62)
	at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:57)
	at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:56)
	at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:36)
	at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.executeTask(EventFiringTaskExecuter.java:77)
	at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:55)
	at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:52)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:416)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:406)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$1.execute(DefaultBuildOperationExecutor.java:165)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:250)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:158)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:102)
	at org.gradle.internal.operations.DelegatingBuildOperationExecutor.call(DelegatingBuildOperationExecutor.java:36)
	at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter.execute(EventFiringTaskExecuter.java:52)
	at org.gradle.execution.plan.LocalTaskNodeExecutor.execute(LocalTaskNodeExecutor.java:41)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:370)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:357)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:350)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:336)
	at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.lambda$run$0(DefaultPlanExecutor.java:127)
	at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.execute(DefaultPlanExecutor.java:191)
	at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.executeNextNode(DefaultPlanExecutor.java:182)
	at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.run(DefaultPlanExecutor.java:124)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
Caused by: java.lang.StackOverflowError
	at kotlin.collections.CollectionsKt__IterablesKt.collectionSizeOrNull(Iterables.kt:33)
	at kotlin.collections.SetsKt___SetsKt.plus(_Sets.kt:119)
	at com.ismaeldivita.changetracker.project.ProjectDependents.getAllDependents(ProjectDependents.kt:16)
	at com.ismaeldivita.changetracker.project.ProjectDependents.getAllDependents(ProjectDependents.kt:16)
	at com.ismaeldivita.changetracker.project.ProjectDependents.getAllDependents(ProjectDependents.kt:16)
	at com.ismaeldivita.changetracker.project.ProjectDependents.getAllDependents(ProjectDependents.kt:16)
	at com.ismaeldivita.changetracker.project.ProjectDependents.getAllDependents(ProjectDependents.kt:16)
	at com.ismaeldivita.changetracker.project.ProjectDependents.getAllDependents(ProjectDependents.kt:16)
	at com.ismaeldivita.changetracker.project.ProjectDependents.getAllDependents(ProjectDependents.kt:16)
```

--- a/README.md
+++ b/README.md
@@ -29,12 +29,14 @@ changeTracker {
     tasks = ['lint','testDebugUnitTest']
     whitelist = [':app']
     blacklist = [':network',':featureA']
+    reevaluate = [':sharedTestModule']
     branch = "master"
 }
 ```
 - `tasks`: List of tasks the plugin will need to create. 
 - `whitelist` (optional): List of modules that should **always** run.
 - `blacklist` (optional): List of modules that should **never** run.
+- `reevaluate` (optional): List of modules that will trigger the task for all modules
 - `branch`: Name of the branch that should compare to the current one to extract the diff.
 
 ## Usage

--- a/README.md
+++ b/README.md
@@ -29,14 +29,15 @@ changeTracker {
     tasks = ['lint','testDebugUnitTest']
     whitelist = [':app']
     blacklist = [':network',':featureA']
-    reevaluate = [':sharedTestModule']
     branch = "master"
 }
 ```
+<!--- Add this above (comment does not work): reevaluate = [':sharedTest'] --->
+
 - `tasks`: List of tasks the plugin will need to create. 
 - `whitelist` (optional): List of modules that should **always** run.
 - `blacklist` (optional): List of modules that should **never** run.
-- `reevaluate` (optional): List of modules that will trigger the task for all modules
+<!--- - `reevaluate` (optional): List of modules that will trigger the task for all modules --->
 - `branch`: Name of the branch that should compare to the current one to extract the diff.
 
 ## Usage
@@ -59,6 +60,23 @@ You can override the default branch used for the comparison when running your co
 ./gradlew testChangedModules -Pbranch=dev
 ```
 
+<!---
+## Cyclic Test Dependencies
+In modular Android projects often testing code exists, which is shared between modules. 
+In such cases, the testing code resides in a module and is used by others modules using
+`testImplementation project(path: ":sharedTest")`. 
+This could lead to cyclic dependencies if the `:sharedTest` depends on another module which uses the `:sharedTest` for tests itself.
+Let's create an example with 2 modules:
+- `:database`: A module which contains code to handle the app's database. Its test requires shared testing code and thus is configured to depend on
+the `:sharedTest` using `testImplementation(project(:sharedTest))`
+- `sharedTest`: Our module with shared test code. As it provides mocks for database testing it depends on `:database`.
+So it has a dependency on `:database` with `implementation(project(:database))`
+
+In this particular example, we created a cyclic dependency if the configuration (`testImplementation`, `androidTestImplementation`, ...) is not considered.
+Because of this, test configurations are not part of the dependency evaluation. 
+However, if you apply changes to `:sharedTest`, `:database` needs to be verified as well. 
+In this case, specify the `:sharedTest` module in the `reevaluate` list to trigger your verification tasks for all modules. 
+--->
 
 <br clear="right"/>
 

--- a/change_tracker/src/main/kotlin/com/ismaeldivita/changetracker/ChangeTrackerExtension.kt
+++ b/change_tracker/src/main/kotlin/com/ismaeldivita/changetracker/ChangeTrackerExtension.kt
@@ -7,6 +7,7 @@ open class ChangeTrackerExtension : Serializable {
     var tasks = emptySet<String>()
     var whitelist = emptySet<String>()
     var blacklist = emptySet<String>()
+    var reevaluate = emptySet<String>()
     var branch = "master"
 
 }

--- a/change_tracker/src/main/kotlin/com/ismaeldivita/changetracker/ChangedModulesTask.kt
+++ b/change_tracker/src/main/kotlin/com/ismaeldivita/changetracker/ChangedModulesTask.kt
@@ -12,6 +12,7 @@ open class ChangedModulesTask : DefaultTask() {
 
     private val whitelist by lazy { getProjectsByName(changeTrackerExtension.whitelist) }
     private val blacklist by lazy { getProjectsByName(changeTrackerExtension.blacklist) }
+    private val reevaluate by lazy { getProjectsByName(changeTrackerExtension.reevaluate) }
     private val branch by lazy { getProperty<String>("branch") ?: changeTrackerExtension.branch }
 
     override fun getGroup(): String? = CHANGED_TRACKER_GROUP_NAME
@@ -26,6 +27,7 @@ open class ChangedModulesTask : DefaultTask() {
 
         val result: MutableSet<Project> = when {
             changedProjects.contains(rootProject) -> rootProject.subprojects
+            reevaluate.any(changedProjects::contains) -> rootProject.subprojects
             changedProjects.contains(null) -> rootProject.subprojects
             else -> changedProjects
                 .filterNotNull()

--- a/change_tracker/src/main/kotlin/com/ismaeldivita/changetracker/project/ProjectDependents.kt
+++ b/change_tracker/src/main/kotlin/com/ismaeldivita/changetracker/project/ProjectDependents.kt
@@ -18,12 +18,14 @@ class ProjectDependents(rootProject: Project) {
 
     private fun Map<Project, MutableSet<Project>>.traverseDependencies() {
         forEach { subProject, _ ->
-            subProject.configurations.forEach { config ->
-                config
-                    .dependencies
-                    .filterIsInstance<ProjectDependency>()
-                    .forEach { getValue(it.dependencyProject).add(subProject) }
-            }
+            subProject.configurations
+                .filter { !it.name.toLowerCase().contains("test") }
+                .forEach { config ->
+                    config
+                        .dependencies
+                        .filterIsInstance<ProjectDependency>()
+                        .forEach { getValue(it.dependencyProject).add(subProject) }
+                }
         }
     }
 

